lib/shopify-cli/contextual_resolver.rb

```diff
@@ -0,0 +1,44 @@
+require 'cli/kit'
+
+module ShopifyCli
+  class ContextualResolver < CLI::Kit::Resolver
+    def call(args)
+      args = args.dup
+      return super(args) unless args.first
+      if args.first.include?('-h') || args.first.include?('--help')
+        help = Commands::Help
+        help.ctx = Context.new
+        help.call([], nil)
+        raise ShopifyCli::AbortSilent
+      else
+        command_name = args.shift
+
+        command, resolved_name = @command_registry.lookup_command(command_name)
+
+        if command.nil?
+          command_not_found(command_name)
+          raise CLI::Kit::AbortSilent # Already output message
+        end
+
+        unless command.available?
+          CLI::UI::Frame.open("Command not available", color: :red, timing: false) do
+            $stderr.puts(CLI::UI.fmt("{{command:#{@tool_name} #{command_name}}} not available here"))
+          end
+          raise CLI::Kit::AbortSilent
+        end
+
+        if command.app_type?
+          project = Project.current
+          cmd = command.app_type_lookup[command][project.app_type_id]
+          unless cmd
+            $stderr.puts(CLI::UI.fmt("{{command:#{command_name}}} not supported in #{project.app_type_id} apps}}"))
+            raise CLI::Kit::AbortSilent
+          end
+          return [cmd, resolved_name, args]
+        end
```

Comment:
I really think here we should have a generic design for when command isn't specific to an app project.  I could be missing something...., but I think `app_type` only needs to be known within the context of an app project, and thus only those contextual command within this context should know what that means.
Let's use `serve` command as an example, Commands::Serve knows about NodeServe and RailsServe.  What if a ContextualCommand has a method `translate_command` which maps to the actual command.  
1. So, on line 17, we essentially calls `Commands::Serve.translate_command`.  Commands::Serve knows the context it's in, and knows that it's inside a rails app project, and then calls `Commands.register(:Serve, 'serve', 'shopify-cli/commands/serve/rails')`.  Once the actual command is registered, we could preserve the existing logic...I think?

2. For this to work, each ContexualCommand has its own `determine_context` method, so `Commands::Serve` knows how to read .yml file to determine its context.  We could have a generic way of determining the path for the actual command class.  So, let's say the path we want is: `shopify-app-cli/commands/serve/app/rails`.  We could have an AppProjectResolver that understands `app_type`.  Similarly, eventually, we will have a ScriptProjectResolver that understands `script_type`.  Basically, these 2 resolvers could have a method `get_command_path` which returns the path for the actual command class.

I was thinking about this when we did our demo a while back.  I think we all want a more generic and clean way for others to plug into the CLI framework here.  Let me know if any of this make sense?

Reply:
I think I understand but I'm not sure whether it would be simpler. I would love to see a prototype.

Reply:
I think I should start integrating with your approach first before testing out other ideas.  It might be easier to find the optimal solution when we see how Scripts commands integrates.  I would like to hear about how you intend to solve the `deploy` command case above though.